Mmacferrin pull request: "devel" <-- "mmacferrin"

mmacferrin opened the pull request:
Per @iparask's suggestion in Ticket #22, to re-merge the branches and develop more iteratively from there.

mmacferrin committed:
Adding functionality within the fast_IMAS_IPOL executable in order to read and subset GeoTIFF files directly.

mmacferrin committed:
Signed-off-by: Mike MacFerrin <michael.macferrin@colorado.edu>

mmacferrin committed:
Signed-off-by: Mike MacFerrin <michael.macferrin@colorado.edu>

mmacferrin committed:
Make sure ASIFT v2.1 code is updated, with GeoTIFF images.
Update the README.txt file in this directory.

Signed-off-by: Mike MacFerrin <michael.macferrin@colorado.edu>

mmacferrin committed:
Adding some outputs of a bit of performance analysis of the ASIFT Executable. Will update later as the run continues.

I will add details about this to the "Boulder" Wiki page.

mmacferrin committed:
Merge branch 'mmacferrin' of https://github.com/iceberg-project/ASIFT into mmacferrin

mmacferrin committed:
Merge branch 'mmacferrin' of https://github.com/iceberg-project/ASIFT into mmacferrin

mmacferrin committed:
Merge branch 'mmacferrin' of https://github.com/iceberg-project/ASIFT into mmacferrin

mmacferrin committed:
Updating the plot.

mmacferrin committed:
Updated plots for performance plotting

Signed-off-by: Mike MacFerrin <michael.macferrin@colorado.edu>

mmacferrin committed:
Mike MacFerrin: Committed code for performance plotting. It requires two WV01 images that I will send Dropbox links to privately over email.

mmacferrin committed:
Uploading "README.txt" and "WorldView_URLs.txt" to the "Performance Plotting" directory.

mmacferrin committed:
Updated "WorldView_URLs.txt" to contain the actual URLs.

mmacferrin committed:
Code for 3 Phases, including READMEs

Not all functionality has been thoroughly tested, some are just stub functions still to be implemented, and some of the formatting on the README.md files needs to be fixed, but it's a major update.

mmacferrin committed:
Cleaning up a few files

mmacferrin committed:
Updating formatting

mmacferrin committed:
Update README.md

mmacferrin committed:
Update README.md

mmacferrin committed:
Readme formatting fixes

mmacferrin committed:
ASIFT Ticket #21 - Got rid of README.docx (left README.md markup files for future edits).
ASIFT Ticket #17 - Reorganized directories (/src/ and /doc/) in the way that Aymen suggested.
ASIFT Ticket #19 (partial) - Renamed code files to get rid of "CamelCase" (convert to all lowercase: "no_camel_case"). Unfinished because I still need to update the Python source code to relfect those filename changes.

(Amended): Tweaked the directory structure to how Aymen suggested.
src -|
 |- entk_scripts
 |- phase_1_...
 |- phase_2_...
 |- ...
doc -|
 |- use _case_description

mmacferrin committed:
Got rid of "ASIFT_Scratch" directory. Updated primary README.md file with simple instructinos for setting up a working-scratch directory on the target machine.

AymenFJA commented on src/PHASE_3_RANSAC_FILTERING/ransac_filter.py:
```diff
@@ -0,0 +1,190 @@
+from osgeo import gdal, gdalconst
+
+def find_nodata_matches(keypoints_array, img_filename, img_nodata=None):
+    '''Take the keypoints given for this image, find the nodata value, and search the image to find all points that are adjacent to nodata values.
```
I would suggest keeping the comments short as much as possible.
```suggestion
    '''Take the keypoints given for this image, find the nodata value, and search the image to find all points that are adjacent to nodata values.
```

mmacferrin replied:
Aymen, shorter comments are fine with me. I was just putting together a doc-string for the function to use for documentation later, but obviously it can be modified however y'all feel is the best way to present it.

iparask replied:
I would like to correct that a bit. @mmacferrin I agree the doc string should be there to be used in the documentation later, and thank you for being ahead of the curve there! I think what @AymenFJA wanted to say is try and keep each line around 80 to 100 characters.

It usually makes the code a bit easier to read!

mmacferrin replied:
Ah yes, of course. And I agree. I have a bad habit of using my widescreen monitor and not paying much attention to the comment width, which I always notice later when opening it on my laptop. I can take care to fix that and shorten the comment widths. 80 chars is pretty standard.

AymenFJA commented on src/PHASE_3_RANSAC_FILTERING/ransac_filter.py:
```diff
@@ -0,0 +1,190 @@
+#######################################################################################=
+
+
+from Matches_Reader import ASIFT_Matches
```
capitalization issue, file names in the directory are lowercase while here it is uppercase not matching the file names.

mmacferrin replied:
You're right, I'm fixing the capitalization issue in a few of the files, as well as the capitalization of the PHASE_... directories (turn them to lowercase "phase_...."). I can either re-commit those changes to my branch. If I do, would they make it into the pull request? (I'm guessing not.) If there's a best-case workflow for me to make those changes, just let me know.

bspitzbart replied:
You can push new changes to your branch and they will be included in the pull request.

AymenFJA replied:
yes, I agree with @bspitzbart . Also we will keep reviewing the pull request until all the issues are solved then we can move forward and accept it to be merged with devel branch.

mmacferrin replied:
Thumbs up. Just committed some of the small fixes, although admittedly I didn't fully test all the code to ensure bugs didn't creep in. We have a telecon in 5 minutes, I'll test and make bug fixes as needed.

(Aymen caught one with the command-line interface in the ransac_filter.py code, for instance.)

AymenFJA replied:
Thanks, @mmacferrin for the fast responses.

mmacferrin committed:
Updated Workflow Diagram, Code bug fixes

Updated the worlflow diagram to reflect the current state of the code, addressing Ticket #14  (Not quite finished, still need to flesh out Phases 2 & 3 in detail like I did with Phase 1. I will close the ticket when those are finished).

Fixed directory name case errors (Ticket #17), commaned_line error in ransac_filter.py, and name (upper-case) errors in other python files.

mmacferrin committed:
Temporary change to directory names

Capitalization issues causing trouble, renaming folders temporarily to see if Git fixes it. Then will change back and re-commit.

mmacferrin committed:
Folder change revert

Folders renamed back to their old names, hopefully Git picks up the lower-case naming of the directories.

AymenFJA commented on src/ASIFT_Master_Deliverable_1.py:
```diff
@@ -0,0 +1,51 @@
+# Import sub-directory phases
+import PHASE_1_IMAGE_PREPROCESSING.ASIFT_Tile_Images as TILING
+import PHASE_2_KEYPOINT_GENERATION.ASIFT_Scheduler as ASIFT
+import PHASE_3_RANSAC_FILTERING.ASIFT_RANSAC_filter as RANSAC
```
@mmacferrin  Hello Mike, 

I think we should convert these to lowercase as well.

mmacferrin replied:
Agreed, and consistent with the comments below, the code would be better named "process_two_images.py" which describes what it does. I hadn't touched that code in awhile, and forgot to update it with the new directory names in mind.

AymenFJA commented on src/utilities/eprint.py:
```diff
@@ -0,0 +1,14 @@
+from __future__ import print_function
+import sys
+
+def eprint(*args, **kwargs):
```
@mmacferrin , 

why do we need eprint, and where do we use it exactly, thanks?

iparask replied:
I agree here with @AymenFJA . This function just import `print` from future. It makes more sense to have the `from...` part where it is used.

It will also make the transition to python3 later in the year easier.

mmacferrin replied:
I was using it from the asift_executable.py code in Phase 2 (where outputs to stdout and stderr were being handled a bit differently). But if it's easier to just use the print() function in Python 3 (agreed on that) and you don't feel this function is needed at all, feel free to drop it, no worries from me. This could easily be done after the merge, for instance.

iparask replied:
Yes, I agree! Thank you for clarifying

AymenFJA commented on src/phase_3_ransac_filtering/ransac_filter.py:
```diff
@@ -0,0 +1,190 @@
+    # 2. Use the cv2.findHomography call, and tweak the parameters to get a good filter.
+    # See reference info: https://docs.opencv.org/3.0-beta/modules/calib3d/doc/camera_calibration_and_3d_reconstruction.html?highlight=findhomography#findhomography
+
+    H, mask = cv2.findHomography(keys1_ndarray, keys2_ndarray, method=cv2.RANSAC, ransacReprojThreshold=ransac_threshold, maxIters=2000)
```
@mmacferrin , 

in this part, I am getting this error , `H, mask = cv2.findHomography(keys1_ndarray, keys2_ndarray, method=cv2.RANSAC, ransacReprojThreshold=ransac_threshold, maxIters=2000)
TypeError: 'maxIters' is an invalid keyword argument for this function` 

so i fixed it 2 days ago and created a pull request under the branch `fix_ransac` by using the following 
 `H, mask = cv2.findHomography(keys1_ndarray, keys2_ndarray, cv2.RANSAC, ransac_threshold, 2000)` .

iparask commented on src/ASIFT_Master_Deliverable_1.py:
```diff
@@ -0,0 +1,51 @@
+import PHASE_3_RANSAC_FILTERING.ASIFT_RANSAC_filter as RANSAC
+
+
+def process_two_single_images_for_keypoints(source_image, target_image, output_CSV, verbose=True):
```
This is the main function for the pipeline? @AymenFJA, should I think that this function will be eventually an EnTK script?

Also, since this is part of a larger framework, does it make sense naming the file `process_two_images` or something similar?

mmacferrin replied:
Agreed that "process_two_images" is a better name for it. I hadn't touched that code in awhile, and indeed perhaps it could be dropped entirely for now, and replaced with a "process_two_images.py" code shortly.

mmacferrin replied:
Just to add, a rapidly growing number of the issues being brought up in this thread have little to do with merging the branch. Things like implementing CSV read/writes with Pandas, and getting rid of the “eprint” function, etc. They aren’t points I disagree with but are unecessary in this pull request, are more just modification suggestions to be done after a merge.

As for the “Todo:” notes in the code, I’ve made it pretty clear that the code is a work in progress. I know y’all are wanting to push it out as soon as possible, but unlike most the other Use Cases, this wasn't a “mostly working” algorithm at the start of this project that we’re implementing for IceBerg, this is a genuinely difficult, perhaps not even fully feasilble, research question that most my time is being spent trying to get a working implementation of. It didn’t exist at all when the project started. It isn’t just “run ASIFT, filter, merge results and we’re done.”

Y’all wanted me to upload the code base as it is, it’s what I uploaded. Yes I’ll get to the TODO items, but right now getting a best-functioning Phase 2 is my top science priority, and it needs to be.

iparask commented on src/file_locations.py:
```diff
@@ -0,0 +1,23 @@
+ASIFT_Scratch_Dir = os.path.join(currentdir, 'ASIFT_Scratch')
+
+# The location where the scratch directory log sits.
+ASIFT_Scratch_Logfile = os.path.join( ASIFT_Scratch_Dir, "Scratch_Directory_Log.csv" )
```
Are these constants used by multiple files? If yes, it perfectly fine having them here, otherwise wouldn't it make sense to add them directly where they are used?

iparask commented on src/phase_1_image_preprocessing/asift_preprocessing.py:
```diff
@@ -0,0 +1,167 @@
+
+    # Compile the list of ASIFT run cases
+    # TODO: Change this to use Run_Case_Scheduler.py ---
+    ASIFT_run_cases_CSV_data = create_asift_run_cases_csv_data(source_tile_CSV_data, target_tile_CSV_data)
```
This will call a function that does not exist, since it is commented out a few lines later. I would suggest that the function is reintroduced and returns something simple.

iparask commented on src/phase_1_image_preprocessing/scratch_directory_manager.py:
```diff
@@ -0,0 +1,230 @@
+        self.logfile_numpy_dt = None
+        return
+
+    def hash_filename(self, fname, remove_path=True):
```
Is there a specific security issue you are concerned about and you are hashing?

mmacferrin replied:
Not a security issue, no. Just getting around having absurdly-super-long directory names by using a hashed string. Since some of the files (WorldView especially) can have very long names that only differ at the end of the string but have the same (say) initial 100 characters, I opted to use that rather than just the first ~40-ish characters which could cause conflicts. If a different method is preferred to generate unique directory names, I'm not married to that one, it's just one that worked.

iparask replied:
I see! Thank you!

iparask commented on src/phase_1_image_preprocessing/scratch_directory_manager.py:
```diff
@@ -0,0 +1,230 @@
+import numpy
+import argparse
+
+class SCRATCH_Manager:
```
Nicely written class. Can you add a docstring explaining the purpose of the class?

iparask commented on src/phase_1_image_preprocessing/tile_planner.py:
```diff
@@ -0,0 +1,388 @@
+from tile_generator import create_constrast_enhanced_tile
+
+def create_tiles_from_image(img_filename,
+                            csv_file=None,
```
Based on the docstring, this seems similar to Seals tilling. I'm not saying that you shouldn't have it, but out of curiosity what are the major differences you see?

iparask commented on src/phase_1_image_preprocessing/tile_writer.py:
```diff
@@ -0,0 +1,179 @@
+
+#import contrast_enhancer
+
+def write_tile(src_filename,
```
Is this different from the `create_tiles_from_image`?

iparask commented on src/phase_2_keypoint_generation/asift_executable.py:
```diff
@@ -0,0 +1,414 @@
+    args = define_and_parse_arguments()
+
+    ASIFT_exe = ASIFT_Executable(output_dir = args.output_dir,
+                                 im1 =       None if args.im1 == ''       else args.im1,
```
When you do an `add_argument` you can add a default value. It will save you the `if`s here.

iparask commented on src/phase_2_keypoint_generation/asift_scheduler.py:
```diff
@@ -0,0 +1,41 @@
+# In an HPC environment, these cases would be farmed out to individual nodes, and their results gathered after execution is finished.
+
+def schedule_and_run_all_ASIFT_tiles(ASIFT_run_cases_CSV_file, verbose=True):
+    '''Takes the images output in the CSV file, runs them against each other, and compile all the outputs together.'''
```
Is this supposed to be the EnTK script?

iparask commented on src/phase_3_ransac_filtering/draw_matches.py:
```diff
@@ -0,0 +1,224 @@
+
+@author: mmacferrin
+
+ASIFT_Output_generator.py -- describes the ASIFT_Output() function to display matching keypoints in two images.
```
I think this is outdated, right?

iparask commented on src/phase_3_ransac_filtering/matches_reader.py:
```diff
@@ -0,0 +1,104 @@
+
+@author: mmacferrin
+
+ASIFT_Matches_reader.py -- A simple utility for opening and reading ASIFT data_matches.csv files
```
This is outdated, right?

iparask commented on src/utilities/csv_auto_reader.py:
```diff
@@ -0,0 +1,138 @@
+import pickle
+
+def read_simple_csv(csv_file, picklefile=None, print_help=False, verbose=False, string_length_size=None):
+
```
I'm curious, why don't you use a panda dataframes to do this?

iparask replied:
Answered in a previous comment. This is resolved

iparask commented on src/utilities/csv_writer.py:
```diff
@@ -0,0 +1,49 @@
+"""
+import numpy
+
+def write_csv(data, csvname, header=None, verbose=False, print_help=False):
```
The same as the previous one. Pandas dataframes allows categorizing data in a column based approach.

mmacferrin replied:
I'm not against it. I'm dating myself... I'd been using a homegrown simple solution like this since before pandas was mainstream. Feel free to change the implementation to use pandas instead if preferred.

iparask replied:
I see :) I was really curious, because it seems as a lot of work!

iparask reviewed (commented):
Hello @mmacferrin,

As probably said before, good work in commenting everything and trying to have nice coding style. I've left some cosmetic comments before diving in and checking the code in more detail. 

Generally, I do not mind merging in devel files that are mainly ToDOs, but are they going to be filed until the end of the month?

I do not want to release files that are ToDOs. If needed we may have to remove them from devel and move them to a feature branch.

Commit:
fix findHomography phase3_ransac

Commit:
removing the .pyc and .csv files

Commit:
removing another .pyc files

Commit:
cleaning utilities dir

mmacferrin committed:
Merge pull request #29 from iceberg-project/fix/issue_27

Makes sense to me. Merge: fix/issue_27

AymenFJA commented on src/phase_1_image_preprocessing/scratch_directory_manager.py:
```diff
@@ -0,0 +1,230 @@
+        # so we don't accidentally concatenate long filenames added to the database.
+        # Max filename size here, 2048 characters.
+        self.logfile_data = read_simple_csv(self.logfile, string_length_size=2048)
+        self.logfile_numpy_dt = self.logfile_data.dtype
```
@mmacferrin  getting `AttributeError: 'NoneType' object has no attribute 'dtype' `  in this line

AymenFJA commented on src/phase_1_image_preprocessing/scratch_directory_manager.py:
```diff
@@ -0,0 +1,230 @@
+    def lookup(self, fname):
+        '''Simply look up a file name in the logfile and return its full scratch directory path. If it doesn't yet exist, return None.'''
+        if self.logfile_data is None:
+            self.read_logfile()
```
@mmacferrin  getting ` AttributeError: 'NoneType' object has no attribute 'dtype' `

mmacferrin committed:
Changing code so that the ASIFT scratch directory doesn't need to be manually set up first, before first run. Read ASIFT Ticket #30 for more details.

This isn't a "user" decision, but more for middleware software managers. When setting up the code to run first time, you'll want to create a Scratch directory where you want the temp-files written, and put that location in file_locations.py. Otherwise, it'll just create a directory where it currently is located in "file_locations.py," which is among the source-code directories. Up to you where you want it.

mmacferrin committed:
Merge branch 'mmacferrin' of https://github.com/iceberg-project/ASIFT into mmacferrin

AymenFJA committed:
Merge branch 'devel' into mmacferrin
